src/main/java/de/tum/in/www1/artemis/service/connectors/jenkins/JenkinsDockerUrlService.java

```diff
@@ -0,0 +1,84 @@
+package de.tum.in.www1.artemis.service.connectors.jenkins;
+
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Optional;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Service;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import de.tum.in.www1.artemis.domain.VcsRepositoryUrl;
+
+@Profile("jenkins")
+@Service
+public class JenkinsDockerUrlService {
+
+    @Value("${jenkins.docker.ci-url:#{null}}")
+    private Optional<URL> dockerCiUrl;
+
+    @Value("${jenkins.docker.vcs-url:#{null}}")
+    private Optional<URL> dockerVcsUrl;
+
+    private static final Logger log = LoggerFactory.getLogger(JenkinsDockerUrlService.class);
+
+    /**
+     * Returns the url pointing the the docker container hosting the vcs
+     * if the vcs url is set. Returns the original url otherwise.
+     * @param vcsRepositoryUrl the vcs repository url
+     * @return the manipulated url
+     */
+    public VcsRepositoryUrl toDockerVcsUrl(VcsRepositoryUrl vcsRepositoryUrl) {
+        if (dockerVcsUrl.isEmpty()) {
+            return vcsRepositoryUrl;
+        }
+
+        if (vcsRepositoryUrl.getURL() == null) {
+            log.warn("Cannot replace url to docker url {} because the url is null.", dockerVcsUrl);
+            return vcsRepositoryUrl;
+        }
+
+        try {
+            String newDockerUrl = replaceUrl(vcsRepositoryUrl.getURL().toString(), dockerVcsUrl.get());
+            return new VcsRepositoryUrl(newDockerUrl);
+        }
+        catch (MalformedURLException e) {
+            log.warn("Cannot replace url {} to {}: {}. Falling back to original url.", vcsRepositoryUrl, dockerVcsUrl, e.getMessage());
+            return vcsRepositoryUrl;
+        }
+    }
+
+    /**
+     * Returns the url pointing the the docker container of the ci
+     * if the ci url is set. Returns the original url otherwise.
+     * @param ciUrl The url of the ci
+     * @return the manipulated url
+     */
+    public String toDockerCiUrl(String ciUrl) {
+        if (dockerCiUrl.isPresent()) {
```

Comment:
maybe use ternary condition instead and write it in one line

Reply:
I think it's a matter of preference :D

Reply:
it looks cleaner that way but we can keep it like it is